jep/0000/README.adoc

```diff
@@ -0,0 +1,296 @@
+
+=== Design decisions
+
+This section contains a top-level description of concerns we need to address in the design.
+
+==== Fingerprint Storage destinations
+
+The solution should be generic enough in order to support common storage types. The following storage types are supported:
+
+* Local XML-based file system storage (current implementation)
+* Redis backed Fingerprint Storage Plugin
+
+The API does not force a particular ORM mapping for the storages.
+The storage plugins are passed on the `Fingerprint` objects, leaving it up to them to decide how to save the fingerprints.
+
+==== Instance Specificness of Fingerprints
+
+[NOTE]
+====
+In this proposal, fingerprint's unique ID (or ID) is used to refer (as of now) to the MD5 checksum of the fingerprinted file.
+====
+
+We propose a design where it is the responsibility of the plugin to provide instance specific fingerprints to Jenkins core when required for traditional read/write/update.
+This means that fingerprints stored in the storage with the same ID by two different Jenkins instances will not be merged.
+
+The idea behind this is that the external storage will work as a common store for all the fingerprints, but the intended design of current offering isn’t affected.
+
+However, the API includes functionality to retrieve fingerprints from other instances in the case when they are required for tracing purposes.
+This allows for tracing fingerprints across Jenkins instances.
+Implementation wise, this is achieved by plugins ensuring that Fingerprints have an associated Jenkins Instance ID.
+
+==== Local to External Migration
+
+When a plugin is configured and enabled, the fingerprints are transferred from the local storage to the external storage.
+
+This by followed by clearing the local database of fingerprints from Jenkins HOME ONLY if the writes are successful.
+
+==== Cleaning up Fingerprints
+
+The current implementation of `FingerprintCleanupThread` works by periodically iterating over the fingerprints and deleting the ones whose builds and jobs are no longer present in the system.
+
+For an external storage implementation, this creates a problem because the external storage cannot query a Jenkins instance.
+
+This proposal offers the following solution:
+
+* When a build/job is deleted from Jenkins, a subsequent update is performed to the fingerprint to reflect this change
+* An async periodic work is run to check if any fingerprint has no jobs/builds associated with it and delete it. (But, it would ensure that if there is a deletion blocking `FingerprintFacet`, then the fingerprint would not be deleted.)
+
+The above solution is provided to the user as an option, and can be toggled because it may be the case that storing data (i.e. fingerprints) might be cheaper than to do consistent updates in the database.
+
+We do not propose to revamp the `FingerprintCleanupThread` to use the above logic at the moment.
+
+==== External to Local Migration (Nice 2 Have)
```

Comment:
Change "nice 2 have"s to "out of scope for MVP" ... or at the very least, s/2/to/ ;)

Reply:
I'd probably actually move these sections to the bottom so they don't distract from the important stuff.

Reply:
@afalko thanks! Incorporated "out of scope for MVP" :) 
I have these at the end of the Design Decisions section